### PR TITLE
Run CI on self-hosted machine

### LINKED ISSUE #451: Run CI on self-hosted machine

### PR DESCRIPTION
**Pull request by kunxian-xia**
Closes #451.

### CODE REVIEW

File: scripts/ci/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    && tar xzf ./actions-runner-linux-x64-$RUNNER_VERSION.tar.gz
+
+RUN chown -R docker ~docker && cd /home/docker/actions-runner/bin/ \
+    && chmod +x installdependencies.sh \
+    && ./installdependencies.sh
+
+COPY start.sh start.sh
+RUN chmod +x start.sh
+
+# ./config.sh & ./run.sh are not allowed to be run as root
```
[the remark this answers is not in the document]
Comment (kunxian-xia) [reply scripts/ci/Dockerfile]: Use better wordings in https://github.com/scroll-tech/ceno/pull/499/commits/fb050e1e3f1d1392e2e27892eaef1c3eef753002.

**Comment by kunxian-xia** [comment]
> Please add some context in the README that describes what someone needs to do to get changes to these files to be reflected in the running CI. I assume it's not automatic like with changes to eg lints.yml?

@matthiasgoergens  Added in https://github.com/scroll-tech/ceno/pull/499/commits/b64ab0fa7165b46c25ca0567a2be7ab7c17ea569.

**Comment by kunxian-xia** [comment]
The test `scheme::tests::test_rw_lk_expression_combination` is hanging for like 30 mins.

File: scripts/ci/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,31 @@
+# this dockerfile is borrowed from https://baccini-al.medium.com/how-to-containerize-a-github-actions-self-hosted-runner-5994cc08b9fb
+FROM ubuntu:22.04
```
Comment (matthiasgoergens) [inline scripts/ci/Dockerfile:2]: We are already running our CI on Ubuntu 24.04.  No need to downgrade.

**Review by matthiasgoergens — COMMENTED** [review]
How does doing this manually compare to using the official https://github.com/actions/runner/pkgs/container/actions-runner?

File: README.md [hunk 3]
```diff
@@ -36,3 +36,12 @@ cargo clippy
...
 
 Alas, `cargo build` doesn't work. That's a known problem and we're working on it.  Please use `cargo make build` instead for now.
+
+### Setting up self-hosted CI docker container
+
+To set up docker container for CI, you can run the following command:
+
+```sh
+docker build -t ceno-runner scripts/ci/
+docker run -d ceno-runner
```
Comment (matthiasgoergens) [inline README.md:46]: I can execute these docker commands locally, but I don't think that's going to change the container that GitHub actually uses?  Or does it?

Comment (matthiasgoergens) [reply README.md:46]: Can you leave some remarks in the ReadMe about who's responsible for our runners, and how to reach them, when eg something breaks or we want to upgrade?

**Review by matthiasgoergens — CHANGES_REQUESTED** [review]
Please see comments.

File: README.md (hunk 3 above)
Comment (kunxian-xia) [reply README.md:46]: You have to click the  button `add new runner` explicitly.

**Comment by kunxian-xia** [comment]
> https://github.com/scroll-tech/ceno/pull/499#pullrequestreview-2414747031

I

> How does doing this manually compare to using the official https://github.com/actions/runner/pkgs/container/actions-runner?

I don't know how the official container work.

**Review by matthiasgoergens — CHANGES_REQUESTED** [review]
Let's just use the official [image](https://github.com/actions/runner/pkgs/container/actions-runner), instead of cobbling together our own.  Less hassle that way.

File: README.md (hunk 3 above)
Comment (matthiasgoergens) [reply README.md:46]: Ok, makes sense.  But I doubt you are running that on your laptop?

Could you add just a tiny bit of context about what you have to do?  Presumably you contact someone in Scroll to help you get access to a machine or something like that?

**Comment by kunxian-xia** [comment]
The default CI runner provided by GitHub is not powerful enough to run the fibonacci example in #368.

**Comment by naure** [comment]
Have you tried `--max-steps=11539` ?

Or any other number of steps. That makes it as fast as you want. You can extrapolate the timings to the full 11_538_921 steps (e.g. multiply by 1000).

**Review by matthiasgoergens — CHANGES_REQUESTED** [review]
Please see my outstanding comments.

Please don't do even more copy and paste, and please provide some comments for how someone would go and fix the runner.

File: scripts/ci/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,31 @@
+# this dockerfile is borrowed from https://baccini-al.medium.com/how-to-containerize-a-github-actions-self-hosted-runner-5994cc08b9fb
```
[the remark this answers is not in the document]
Comment (matthiasgoergens) [reply scripts/ci/Dockerfile:1]: But please don't copy and paste.

**Review by naure — APPROVED** [review]

**Comment by kunxian-xia** [comment]
Please refer to https://www.notion.so/scrollzkp/Ceno-s-self-hosted-CI-runner-1437792d22af800eba16e4b971e82619?pvs=4 for more detailed instructions on setting up our private CI runner.

**Comment by matthiasgoergens** [comment]
> Please refer to https://www.notion.so/scrollzkp/Ceno-s-self-hosted-CI-runner-1437792d22af800eba16e4b971e82619?pvs=4 for more detailed instructions on setting up our private CI runner.

Please put that where someone can find it.  Put it in the README for example.  No one looks in old reviews when their hair is on fire.

Having said that, this is better than nothing..

**Pull request merged**